### PR TITLE
net: remove pvconn reference from all devif callback

### PR DESCRIPTION
**Pull request by anchao**
## Summary

net: cleanup pvconn reference to avoid confuse
net: remove pvconn reference from all devif callback

Do not use 'pvconn' argument to get the connection pointer since
pvconn is normally NULL for some events like NETDEV_DOWN.
Instead, the connection pointer can be reliably obtained from the
corresponding private pointer.

More reference:
@a-lunev 
https://github.com/apache/incubator-nuttx/pull/5252
https://github.com/apache/incubator-nuttx/pull/5434

## Impact

N/A

## Testing

tcp/udp/usrsock/icmp

### CODE REVIEW

**Comment by anchao** [comment]
@PetervdPerk-NXP 

https://github.com/apache/incubator-nuttx/pull/6740

Could you please review this PR? since the using pvconn from devif callback will have some risk if netdev shutdown, In this PR, I brought back the "clscb" into tcp_conn_s, and compared the RAM consumption of tcp:

```
./tools/configure.sh  spresense/rndis

Without this PR:
$ size nuttx
   text	   data	    bss	    dec	    hex	filename
 324784	    640	  45348	 370772	  5a854	nuttx

With this PR:
$ size nuttx
   text	   data	    bss	    dec	    hex	filename
 324752	    640	  45380	 370772	  5a854	nuttx
```
I do not know if this result is acceptable?

File: net/icmp/icmp_netpoll.c [hunk 1]
```diff
@@ -61,7 +61,6 @@
  ****************************************************************************/
 
 static uint16_t icmp_poll_eventhandler(FAR struct net_driver_s *dev,
-                                       FAR void *pvconn,
```
Comment (xiaoxiang781216) [inline net/icmp/icmp_netpoll.c:64]: update line 52

File: net/icmp/icmp_recvmsg.c [hunk 2]
```diff
@@ -93,8 +93,7 @@ struct icmp_recvfrom_s
  ****************************************************************************/
 
 static uint16_t recvfrom_eventhandler(FAR struct net_driver_s *dev,
-                                  FAR void *pvconn,
```
Comment (xiaoxiang781216) [inline net/icmp/icmp_recvmsg.c:96]: remove line 83

File: net/icmpv6/icmpv6_netpoll.c [hunk 3]
```diff
@@ -61,7 +61,6 @@
  ****************************************************************************/
 
 static uint16_t icmpv6_poll_eventhandler(FAR struct net_driver_s *dev,
-                                         FAR void *pvconn,
```
Comment (xiaoxiang781216) [inline net/icmpv6/icmpv6_netpoll.c:64]: update line 52

File: net/icmpv6/icmpv6_recvmsg.c [hunk 4]
```diff
@@ -96,8 +96,7 @@ struct icmpv6_recvfrom_s
  ****************************************************************************/
 
 static uint16_t recvfrom_eventhandler(FAR struct net_driver_s *dev,
-                                  FAR void *pvconn,
```
Comment (xiaoxiang781216) [inline net/icmpv6/icmpv6_recvmsg.c:99]: remove line 86

File: net/sixlowpan/sixlowpan_send.c [hunk 5]
```diff
@@ -98,7 +98,6 @@ struct sixlowpan_send_s
  ****************************************************************************/
 
 static uint16_t send_eventhandler(FAR struct net_driver_s *dev,
-                                  FAR void *pvconn,
```
Comment (xiaoxiang781216) [inline net/sixlowpan/sixlowpan_send.c:101]: update line 89

File: net/tcp/tcp_connect.c [hunk 6]
```diff
@@ -149,7 +148,6 @@ static void psock_teardown_callbacks(FAR struct tcp_connect_s *pstate,
  *
  * Input Parameters:
  *   dev      The structure of the network driver that reported the event
- *   pvconn   The connection structure associated with the socket
```
Comment (xiaoxiang781216) [inline net/tcp/tcp_connect.c:152]: pvconn->pvpriv

File: net/tcp/tcp_recvfrom.c [hunk 7]
```diff
@@ -363,7 +363,6 @@ static inline void tcp_sender(FAR struct net_driver_s *dev,
  *
  * Input Parameters:
  *   dev      The structure of the network driver that generated the event.
- *   pvconn   The connection structure associated with the socket
```
Comment (xiaoxiang781216) [inline net/tcp/tcp_recvfrom.c:366]: pvconn->pvpriv

File: net/tcp/tcp_send_buffered.c [hunk 8]
```diff
@@ -311,17 +311,8 @@ static inline void send_ipselect(FAR struct net_driver_s *dev,
  ****************************************************************************/
 
 static uint16_t psock_send_eventhandler(FAR struct net_driver_s *dev,
-                                        FAR void *pvconn, FAR void *pvpriv,
```
Comment (xiaoxiang781216) [inline net/tcp/tcp_send_buffered.c:314]: update line 302

File: net/tcp/tcp_send_unbuffered.c [hunk 9]
```diff
@@ -169,17 +169,8 @@ static inline void tcpsend_ipselect(FAR struct net_driver_s *dev,
  ****************************************************************************/
 
 static uint16_t tcpsend_eventhandler(FAR struct net_driver_s *dev,
-                                     FAR void *pvconn,
```
Comment (xiaoxiang781216) [inline net/tcp/tcp_send_unbuffered.c:172]: update line 100

File: net/tcp/tcp_sendfile.c [hunk 10]
```diff
@@ -122,17 +122,8 @@ struct sendfile_s
  ****************************************************************************/
 
 static uint16_t sendfile_eventhandler(FAR struct net_driver_s *dev,
-                                      FAR void *pvconn, FAR void *pvpriv,
```
Comment (xiaoxiang781216) [inline net/tcp/tcp_sendfile.c:125]: update line 113

File: net/udp/udp_recvfrom.c [hunk 11]
```diff
@@ -411,7 +411,6 @@ static void udp_terminate(FAR struct udp_recvfrom_s *pstate, int result)
  *
  * Input Parameters:
  *   dev      The structure of the network driver that generated the event.
- *   pvconn   The connection structure associated with the socket
```
Comment (xiaoxiang781216) [inline net/udp/udp_recvfrom.c:414]: pvconn->pvpriv

Comment (anchao) [reply net/udp/udp_recvfrom.c:414]: Done

File: net/icmpv6/icmpv6_netpoll.c (hunk 3 above)
Comment (anchao) [reply net/icmpv6/icmpv6_netpoll.c:64]: Done

**Review by xiaoxiang781216 — APPROVED** [review]

File: net/tcp/tcp_close.c [hunk 12]
```diff
@@ -306,13 +303,13 @@ static inline int tcp_close_disconnect(FAR struct socket *psock)
...
        conn->tcpstateflags == TCP_LAST_ACK) &&
-      (cb = tcp_callback_alloc(conn)) != NULL)
+      (conn->clscb = tcp_callback_alloc(conn)) != NULL)
     {
       /* Set up to receive TCP data event callbacks */
 
-      cb->flags = (TCP_NEWDATA | TCP_POLL | TCP_DISCONN_EVENTS);
-      cb->event = tcp_close_eventhandler;
-      cb->priv  = cb; /* reference for event handler to free cb */
+      conn->clscb->flags = (TCP_NEWDATA | TCP_POLL | TCP_DISCONN_EVENTS);
```
Comment (pkarashchenko) [inline net/tcp/tcp_close.c:310]:
```suggestion
      conn->clscb->flags = TCP_NEWDATA | TCP_POLL | TCP_DISCONN_EVENTS;
```

**Review by pkarashchenko — APPROVED** [review]
LGTM!

**Comment by fjpanag** [comment]
I am testing this PR, and everything seems to be working correctly.

I did test the snippet that was known to crash the system, and now it seems stable:
```c
	sched_lock();

	close(sd);
	netlib_ifdown("eth0");

	sched_unlock();
```

I will update a few devices with this change, to test long-term stability, and will report back if anything bad is observed.

*Testing also included the fix in #6918.*

File: net/tcp/tcp_close.c (hunk 12 above)
Comment (anchao) [reply net/tcp/tcp_close.c]: Done!

**Comment by xiaoxiang781216** [comment]
@fjpanag do you have the final result?

**Comment by fjpanag** [comment]
@xiaoxiang781216 I have performed some tests, and I used the patch for some time. No issues so far.

Now I have applied the patch on the production firmware and I have left it working. However this will take a lot of time. In a real-world case the issue may appear after days of normal operation.

I suggest to merge this for the moment, and *if* I encounter any problems in the future, I will report back.

**Comment by fjpanag** [comment]
I just had one device crash again.

The original issue of the NULL pointer is indeed resolved, but I believe the new problem is strongly related to the original issue (or maybe this patch).

An assertion fails at `devif_callback.c` at line 85. A callback is double-freed during TCP close.

Here is the call trace:

image


What may be interesting to you is that the second worker was trying to execute `tcp_close_work()` at the same time, but it was blocked due to the network being locked:


image

**Comment by xiaoxiang781216** [comment]
@anchao please take a look.

**Comment by anchao** [comment]
> I just had one device crash again.
@fjpanag 
Thanks a lot for your analysis, I have a fix but can not evaluate the side effect, I am not sure if that can resolved your issue.
https://github.com/apache/incubator-nuttx/pull/6947

**Pull request merged**